File: r/R/region-expression.R
# Boolean region-expression parser (Phase 11 — Feature D).
#
# Translates an expression like "A & B + !C" into a sorted integer vector of
# region bitmasks. Designed to compose with:
#   - render_venn_svg(..., highlight = parse_region_expression(...))
#   - exclusive_items() / intersection_items() / union_items() in user code.
#
# Grammar (precedence: ~ / ! > & > | / +; left-associative):
#   expr   := term ((`|` | `+`) term)*
#   term   := factor (`&` factor)*
#   factor := (`~` | `!`) factor | `(` expr `)` | letter
#   letter := A | B | ... | I (uppercase, ASCII)
#
# Implementation: hand-rolled tokenizer + shunting-yard to RPN + RPN eval
# over the bitmask space {1..2^n - 1}.

#' @noRd
# Tokenize an expression string into a list of tokens.
# Tokens: list(type = "letter"|"op"|"paren", value = string).
.tokenize_region_expr <- function(expr) {
    if (!nzchar(trimws(expr))) {
        stop("region expression is empty", call. = FALSE)
    }
    chars <- strsplit(expr, "", fixed = TRUE)[[1L]]
    tokens <- list()
    for (ch in chars) {
        if (grepl("^[ \t\n]$", ch)) next
        if (ch %in% c("&", "|", "+", "~", "!")) {
            tokens[[length(tokens) + 1L]] <- list(type = "op", value = ch)
        } else if (ch %in% c("(", ")")) {
            tokens[[length(tokens) + 1L]] <- list(type = "paren", value = ch)
        } else if (grepl("^[A-I]$", ch)) {
            tokens[[length(tokens) + 1L]] <- list(type = "letter", value = ch)
        } else {
            stop(sprintf("Unrecognised character in region expression: %s",
                          sQuote(ch)), call. = FALSE)
        }
    }
    tokens
}

#' @noRd
# Operator precedence + associativity for shunting-yard.
# Higher number = higher precedence. All are left-associative except `~`/`!`.
.op_precedence <- function(op) {
    switch(op,
        "~" = 4L, "!" = 4L,
        "&" = 3L,
        "|" = 2L, "+" = 2L,
        0L)
}

#' @noRd
.is_unary <- function(op) op %in% c("~", "!")

#' @noRd
# Convert infix token list to RPN (Reverse Polish Notation) token list via
# Dijkstra's shunting-yard. Throws on unbalanced parens or malformed input.
.shunting_yard <- function(tokens) {
    output <- list()
    op_stack <- list()
    expect_operand <- TRUE
    for (i in seq_along(tokens)) {
        t <- tokens[[i]]
        if (t$type == "letter") {
            if (!expect_operand) {
                stop("Malformed region expression: unexpected letter", call. = FALSE)
            }
            output[[length(output) + 1L]] <- t
            expect_operand <- FALSE
        } else if (t$type == "op") {
            if (.is_unary(t$value)) {
                if (!expect_operand) {
                    stop("Malformed region expression: unexpected unary op", call. = FALSE)
                }
                op_stack[[length(op_stack) + 1L]] <- t
                # still expecting operand after unary
            } else {
                if (expect_operand) {
                    stop("Malformed region expression: binary op without left operand",
                         call. = FALSE)
                }
                while (length(op_stack) > 0L) {
                    top <- op_stack[[length(op_stack)]]
                    if (top$type == "op" &&
                        .op_precedence(top$value) >= .op_precedence(t$value)) {
                        output[[length(output) + 1L]] <- top
                        op_stack[[length(op_stack)]] <- NULL
                    } else {
                        break
                    }
                }
                op_stack[[length(op_stack) + 1L]] <- t
                expect_operand <- TRUE
            }
        } else if (t$type == "paren" && t$value == "(") {
            op_stack[[length(op_stack) + 1L]] <- t
            expect_operand <- TRUE
        } else if (t$type == "paren" && t$value == ")") {
            matched <- FALSE
            while (length(op_stack) > 0L) {
                top <- op_stack[[length(op_stack)]]
                op_stack[[length(op_stack)]] <- NULL
                if (top$type == "paren" && top$value == "(") {
                    matched <- TRUE
                    break
                }
                output[[length(output) + 1L]] <- top
            }
            if (!matched) {
                stop("Parenthesis mismatch in region expression", call. = FALSE)
            }
            expect_operand <- FALSE
        }
    }
    while (length(op_stack) > 0L) {
        top <- op_stack[[length(op_stack)]]
        op_stack[[length(op_stack)]] <- NULL
        if (top$type == "paren") {
            stop("Parenthesis mismatch in region expression", call. = FALSE)
        }
        output[[length(output) + 1L]] <- top
    }
    if (expect_operand && length(output) > 0L) {
        stop("Malformed region expression: trailing operator", call. = FALSE)
    }
    output
}

#' @noRd
# Evaluate an RPN token list against the candidate bitmask universe
# {1..2^n - 1}. Each token is interpreted as a set-of-masks; binary ops are
# set operations (intersection / union); unary ~ / ! is complement against
# the candidate universe.
.eval_rpn <- function(rpn, n_sets) {
    universe <- seq_len(bitwShiftL(1L, n_sets) - 1L)
    stack <- list()
    for (i in seq_along(rpn)) {
        t <- rpn[[i]]
        if (t$type == "letter") {
            letter <- t$value
            bit_pos <- match(letter, strsplit(.LETTERS_VDL, "", fixed = TRUE)[[1L]]) - 1L
            if (is.na(bit_pos) || bit_pos >= n_sets) {
                stop(sprintf("Letter %s is out of range for n_sets = %d",
                              sQuote(letter), n_sets), call. = FALSE)
            }
            bit <- bitwShiftL(1L, bit_pos)
            stack[[length(stack) + 1L]] <- universe[bitwAnd(universe, bit) != 0L]
        } else if (t$type == "op") {
            op <- t$value
            if (.is_unary(op)) {
                if (length(stack) < 1L) stop("Malformed expression: unary op with empty stack", call. = FALSE)
                a <- stack[[length(stack)]]; stack[[length(stack)]] <- NULL
                stack[[length(stack) + 1L]] <- setdiff(universe, a)
            } else {
                if (length(stack) < 2L) stop("Malformed expression: binary op with insufficient operands", call. = FALSE)
                b <- stack[[length(stack)]]; stack[[length(stack)]] <- NULL
                a <- stack[[length(stack)]]; stack[[length(stack)]] <- NULL
                result <- switch(op,
                    "&" = intersect(a, b),
                    "|" = union(a, b),
                    "+" = union(a, b),
                    stop(sprintf("Unknown operator %s", sQuote(op)), call. = FALSE))
                stack[[length(stack) + 1L]] <- result
            }
        }
    }
    if (length(stack) != 1L) {
        stop("Malformed region expression: stack imbalance after evaluation",
             call. = FALSE)
    }
    sort(as.integer(stack[[1L]]))
}

#' Parse a Boolean region expression into bitmasks
#'
#' Translates an expression like `"A & B + !C"` into a sorted integer vector
#' of region bitmasks (each `1..2^n_sets - 1`). The output is intended for
#' composition with [render_venn_svg()] (`highlight = ...`) and with the
#' region-accessor family in user code.
#'
#' Grammar:
#' \itemize{
#'   \item `A`, `B`, ..., `I` — set atoms (uppercase ASCII).
#'   \item `&` — intersection.
#'   \item `|` or `+` — union.
#'   \item `~` or `!` — complement (unary, against `1..2^n_sets - 1`).
#'   \item `(`, `)` — grouping.
#' }
#' Precedence (highest first): `~ !`, `&`, `| +`. Binary operators are
#' left-associative.
#'
#' @param expr Character scalar; a Boolean expression in the grammar above.
#' @param n_sets Integer; number of sets in the diagram (2..9).
#' @return Sorted integer vector of region bitmasks. Empty integer vector
#'   when the expression is satisfiable by no region (e.g. `"A & ~A"`).
#' @examples
#' parse_region_expression("A & B", n_sets = 3L)        # c(3L, 7L)
#' parse_region_expression("A + B", n_sets = 3L)        # c(1, 2, 3, 5, 6, 7)
#' parse_region_expression("A & ~B", n_sets = 3L)       # c(1L, 5L)
#' parse_region_expression("(A | B) & C", n_sets = 3L)  # c(5L, 6L, 7L)
#' @export
parse_region_expression <- function(expr, n_sets) {
    if (!is.character(expr) || length(expr) != 1L) {
        stop("`expr` must be a character scalar.", call. = FALSE)
    }
    if (!is.numeric(n_sets) || length(n_sets) != 1L ||
        n_sets < 2L || n_sets > 9L) {
        stop("`n_sets` must be an integer in 2..9.", call. = FALSE)
    }
    n_sets <- as.integer(n_sets)
    tokens <- .tokenize_region_expr(expr)
    rpn <- .shunting_yard(tokens)
    .eval_rpn(rpn, n_sets)
}

File: python/src/venn_diagram_lab/region_expression.py
"""Boolean region-expression parser.

Translates an expression like ``"A & B + !C"`` into a sorted list of
region bitmasks (each ``1..2^n_sets - 1``). Designed to compose with
:func:`venn_diagram_lab.render.svg.render_venn_svg` (``highlight=``
argument) and with the region-accessor family in user code.

Grammar (precedence: ``~`` / ``!`` > ``&`` > ``|`` / ``+``;
left-associative for binary operators):

* atom: ``A``, ``B``, ..., ``I`` (uppercase ASCII)
* unary: ``~`` or ``!`` (complement, against ``1..2^n_sets - 1``)
* binary: ``&`` (intersection), ``|`` or ``+`` (union)
* grouping: ``(``, ``)``

Implementation: hand-rolled tokenizer + shunting-yard to RPN + RPN
evaluation over the bitmask space ``{1..2^n_sets - 1}``.
"""

from __future__ import annotations

from dataclasses import dataclass

__all__ = ["parse_region_expression"]

_LETTERS = "ABCDEFGHI"

# Bounds on `n_sets` accepted by the parser. Matches the 2..9 set diagram
# catalogue supported by the rest of the package.
_MIN_N_SETS = 2
_MAX_N_SETS = 9
_BINARY_ARITY = 2

# Operator precedence. Higher = binds tighter. Unary `~`/`!` is highest.
_PRECEDENCE = {"~": 4, "!": 4, "&": 3, "|": 2, "+": 2}


@dataclass(frozen=True)
class _Token:
    type: str  # "letter" | "op" | "paren"
    value: str


def _tokenize(expr: str) -> list[_Token]:
    if not expr.strip():
        raise ValueError("region expression is empty")
    tokens: list[_Token] = []
    for ch in expr:
        if ch in (" ", "\t", "\n"):
            continue
        if ch in ("&", "|", "+", "~", "!"):
            tokens.append(_Token("op", ch))
        elif ch in ("(", ")"):
            tokens.append(_Token("paren", ch))
        elif ch in _LETTERS:
            tokens.append(_Token("letter", ch))
        else:
            raise ValueError(f"Unrecognised character in region expression: {ch!r}")
    return tokens


def _is_unary(op: str) -> bool:
    return op in ("~", "!")


def _handle_op_token(
    t: _Token,
    output: list[_Token],
    op_stack: list[_Token],
    expect_operand: bool,
) -> bool:
    """Process an operator token; return the new ``expect_operand`` flag."""
    if _is_unary(t.value):
        if not expect_operand:
            raise ValueError("Malformed region expression: unexpected unary op")
        op_stack.append(t)
        return True  # still expecting operand
    if expect_operand:
        raise ValueError(
            "Malformed region expression: binary op without left operand"
        )
    while (
        op_stack
        and op_stack[-1].type == "op"
        and _PRECEDENCE[op_stack[-1].value] >= _PRECEDENCE[t.value]
    ):
        output.append(op_stack.pop())
    op_stack.append(t)
    return True


def _handle_close_paren(output: list[_Token], op_stack: list[_Token]) -> None:
    """Drain the operator stack down to the matching ``(``."""
    while op_stack:
        top = op_stack.pop()
        if top.type == "paren" and top.value == "(":
            return
        output.append(top)
    raise ValueError("Parenthesis mismatch in region expression")


def _shunting_yard(tokens: list[_Token]) -> list[_Token]:
    """Convert infix tokens to RPN via Dijkstra's algorithm."""
    output: list[_Token] = []
    op_stack: list[_Token] = []
    expect_operand = True
    for t in tokens:
        if t.type == "letter":
            if not expect_operand:
                raise ValueError("Malformed region expression: unexpected letter")
            output.append(t)
            expect_operand = False
        elif t.type == "op":
            expect_operand = _handle_op_token(t, output, op_stack, expect_operand)
        elif t.value == "(":
            op_stack.append(t)
            expect_operand = True
        else:  # t.value == ")"
            _handle_close_paren(output, op_stack)
            expect_operand = False
    while op_stack:
        top = op_stack.pop()
        if top.type == "paren":
            raise ValueError("Parenthesis mismatch in region expression")
        output.append(top)
    if expect_operand and output:
        raise ValueError("Malformed region expression: trailing operator")
    return output


def _eval_rpn(rpn: list[_Token], n_sets: int) -> list[int]:
    """Evaluate the RPN token list over the bitmask universe ``1..2^n_sets - 1``."""
    universe = set(range(1, 1 << n_sets))
    stack: list[set[int]] = []
    for t in rpn:
        if t.type == "letter":
            bit_pos = _LETTERS.index(t.value)
            if bit_pos >= n_sets:
                raise ValueError(
                    f"Letter {t.value!r} is out of range for n_sets = {n_sets}"
                )
            bit = 1 << bit_pos
            stack.append({m for m in universe if m & bit})
        elif t.type == "op":
            op = t.value
            if _is_unary(op):
                if not stack:
                    raise ValueError(
                        "Malformed expression: unary op with empty stack"
                    )
                a = stack.pop()
                stack.append(universe - a)
            else:
                if len(stack) < _BINARY_ARITY:
                    raise ValueError(
                        "Malformed expression: binary op with insufficient operands"
                    )
                b = stack.pop()
                a = stack.pop()
                if op == "&":
                    stack.append(a & b)
                elif op in ("|", "+"):
                    stack.append(a | b)
                else:
                    raise ValueError(f"Unknown operator {op!r}")
    if len(stack) != 1:
        raise ValueError("Malformed region expression: stack imbalance")
    return sorted(stack[0])


def parse_region_expression(expr: str, n_sets: int) -> list[int]:
    """Parse a Boolean region expression into a sorted list of bitmasks.

    Parameters
    ----------
    expr : str
        Boolean expression. Grammar: ``A..I`` atoms, ``&`` intersection,
        ``|`` or ``+`` union, ``~`` or ``!`` complement, parentheses for
        grouping. Precedence: unary > ``&`` > union (left-associative).
    n_sets : int
        Number of sets in the diagram (2..9). Sets the universe of valid
        bitmasks to ``1..2^n_sets - 1``.

    Returns
    -------
    list[int]
        Sorted list of region bitmasks that satisfy the expression.
        Empty list when no region satisfies (e.g. ``"A & ~A"``).

    Raises
    ------
    ValueError
        On empty expression, unbalanced parentheses, trailing operator,
        unknown character, or an atom outside the ``A..letter[n_sets-1]``
        range.
    """
    if not isinstance(expr, str):
        raise ValueError("`expr` must be a string.")
    if not isinstance(n_sets, int) or n_sets < _MIN_N_SETS or n_sets > _MAX_N_SETS:
        raise ValueError("`n_sets` must be an integer in 2..9.")
    tokens = _tokenize(expr)
    rpn = _shunting_yard(tokens)
    return _eval_rpn(rpn, n_sets)
